src/Infrastructure/Repositories/LocalFilesRepository.php

```diff
@@ -42,84 +42,107 @@ public function getDefaultDirectory(): string
+        $files = array_filter($paths, fn (array $path): bool => $path['is_file']);
+
+        return $this->files = array_merge(
+            $this->filesWithinDirectories($directories, $this->exclude),
+            $this->filesAtPaths($files)
+        );
     }
 
     public function within(array $paths, array $exclude = []): FilesRepository
     {
-        foreach ($paths as $path) {
-            $pathInfo = pathinfo($path);
-
-            if (! is_dir($path) && is_file($path)) {
-                $this->fileList['dirname'][] = $pathInfo['dirname'];
-                $this->fileList['basename'][] = $pathInfo['basename'];
-                $this->fileList['full_path'][] = $pathInfo['dirname'] . DIRECTORY_SEPARATOR . $pathInfo['basename'];
-            } else {
-                $this->directoryList[] = $pathInfo['dirname'] . DIRECTORY_SEPARATOR . $pathInfo['basename'];
-            }
-        }
-
-        $directoryFiles = $this->directoryList === null ? [] : $this->getDirectoryFiles($exclude);
-        $singleFiles = $this->fileList === null ? [] : $this->getSingleFiles();
-
-        $this->files = array_merge($directoryFiles, $singleFiles);
+        $this->paths = $paths;
+        $this->exclude = $exclude;
 
         return $this;
     }
 
     /**
+     * @param array<string> $directories
      * @param array<string> $exclude
      *
      * @return array<string, \Symfony\Component\Finder\SplFileInfo>
      */
-    private function getDirectoryFiles(array $exclude = []): array
+    private function filesWithinDirectories(array $directories, array $exclude = []): array
     {
-        $this->finder = Finder::create()
+        if (count($directories) === 0) {
+            return [];
+        }
+
+        $directories = array_column($directories, 'full_path');
+
+        $finder = (clone $this->finder)
             ->files()
             ->name(['*.php'])
             ->exclude(['vendor', 'tests', 'Tests', 'test', 'Test'])
             ->notName(['*.blade.php'])
             ->ignoreUnreadableDirs()
-            ->in($this->directoryList ?? [])
+            ->in($directories)
             ->notPath($exclude);
 
         foreach ($exclude as $value) {
             if (substr($value, -4) === '.php') {
-                $this->finder->notName($value);
+                $finder->notName($value);
             }
         }
 
-        return $this->getFilesList();
+        return $this->getFilesList($finder);
     }
 
     /**
+     * @param array<string> $paths
+     *
      * @return array<string, \Symfony\Component\Finder\SplFileInfo>
      */
-    private function getSingleFiles(): array
+    private function filesAtPaths(array $paths): array
     {
-        $this->finder = Finder::create()
-            ->in($this->fileList['dirname'] ?? [])
-            ->name($this->fileList['basename'] ?? '')
+        if (count($paths) === 0) {
+            return [];
+        }
+
+        $dirname = array_column($paths, 'dirname');
```

Comment:
maybe we can add an array_unique around ? 

If I call phpinsights like this 

```bash
bin/phpinsights analyse src/Domain/Analyser.php src/Domain/Runner.php
```

the dirname $PWD/src/Domain/ will be present twice. 

Don't know if it can impact finder performance